cores/esp8266/reboot_uart_dwnld.cpp

```diff
@@ -0,0 +1,112 @@
+#include "reboot_uart_dwnld.h"
+#include <stdnoreturn.h>
+#include <ets_sys.h>
+#include <user_interface.h>
+#include <spi_flash.h>
+#include <esp8266_undocumented.h>
+
+void ICACHE_RAM_ATTR boot_from_something_uart_dwnld(void (**user_start_ptr)())
+{
+	/* simplified for following condition
+	 * (GPI >> 0x10 & 0x07) == 1 and
+	 * (GPI >> 0x1D & 0x07) == 6
+	 */
+
+	const uint32_t uart_no = 0;
+	const uint16_t divlatch = uart_baudrate_detect(uart_no, 0);
+	uart_div_modify(uart_no, divlatch);
+	UartDwnLdProc((uint8_t*)0x3fffa000, 0x2000, user_start_ptr);
+}
+
+[[noreturn]] void ICACHE_RAM_ATTR main_uart_dwnld()
+{
+	// TODO may be it is part of _start()
+	// at least ets_intr_lock() is called in system_restart_local()
+	// Therefore unlock it here again to enable UART Rx IRQ
+	ets_intr_unlock();
+
+	uartAttach();
+	Uart_Init(0);
+	ets_install_uart_printf(0);
+
+	boot_from_something_uart_dwnld(&user_start_fptr);
+
+	// Call _ResetVector() instead of original implementation
+	// Therefore esptool.py stub is not yet supported
+	/* moved from system_restart_core_uart_dwnld() to here
+	 * Without this bit clearing the _ResetVector() would run into
+	 * an exception
+	 */
+	CLEAR_PERI_REG_MASK(PERIPHS_DPORT_24, 0x18);
+	_ResetVector();
+
+	// TODO calling user_start_fptr() fails with an exception
+	// Fatal exception (0):
+	// epc1=0x4010f45c, epc2=0x00000000, epc3=0x00000000, excvaddr=0x00000000, depc=0x00000000
+	// May be execution flag is not set for memory region
+	//if (user_start_fptr == NULL) {
+	//	if (boot_from_flash() != 0) {
+	//		ets_printf("%s %s \n", "ets_main.c", "181");
+	//		while (true);
+	//	}
+	//}
+	//
+	//_xtos_set_exception_handler(EXCCAUSE_UNALIGNED, window_spill_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_ILLEGAL, print_fatal_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_INSTR_ERROR, print_fatal_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_LOAD_STORE_ERROR, print_fatal_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_LOAD_PROHIBITED, print_fatal_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_STORE_PROHIBITED, print_fatal_exc_handler);
+	//_xtos_set_exception_handler(EXCCAUSE_PRIVILEGED, print_fatal_exc_handler);
+	//
+	///* Moved from system_restart_core_uart_dwnld(). Not sure if it is required */
+	//Cache_Read_Disable();
+	//CLEAR_PERI_REG_MASK(PERIPHS_DPORT_24, 0x18);
+	//
+	//if (user_start_fptr) {
+	//	user_start_fptr();
+	//}
+	//
+	//ets_printf("user code done\n");
+	//ets_run();
+}
+
+[[noreturn]] void ICACHE_RAM_ATTR system_restart_core_uart_dwnld()
+{
+	Wait_SPI_Idle(flashchip);
+
+	// TODO exception when calling uart_div_modify()
+	//Cache_Read_Disable();
+	//CLEAR_PERI_REG_MASK(PERIPHS_DPORT_24, 0x18);
+	main_uart_dwnld();
+}
+
+[[noreturn]] void system_restart_local_uart_dwnld()
```

Comment:
This is the only bit of code usable out of this file, so can we make the other functions `static` to minimize namespace pollution?

Reply:
I have reduced it to two methods. Unfortunately when I add static to `esp8266UartDownloadMode()` it does not work anymore. I guess this function will then inlined and there not located in IRAM.